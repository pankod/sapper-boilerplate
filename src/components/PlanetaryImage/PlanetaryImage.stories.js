import PlanetaryImage from "./index";
import { imageUrl, imageType, imageTitle, copyright, photoOfTheDay } from "@Store";

const apod = {
    url:"https://apod.nasa.gov/apod/image/1912/M20_volskiy1024.jpg",
    title:"Messier 20 and 21",
    type:"apod",
    copyright:"Stanislav Volskiy"
};

const setImage = image => {
    imageUrl.set(image.url);
    imageTitle.set(image.title);
    imageType.set(image.type);
    copyright.set(image.copyright);
};

photoOfTheDay.set(apod);
setImage(apod);

export default { title: "PlanetaryImage" }

export const basic = () => ({
    Component: PlanetaryImage,
    props:{
        setCurrentImage: setImage
    }
});
