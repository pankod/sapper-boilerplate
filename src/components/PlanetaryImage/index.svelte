<script>
    // #region Imports
    import {
        imageUrl,
        imageTitle,
        imageType,
        copyright,
        photoOfTheDay,
        searchTerm,
        searchResults,
    } from "@Store";
    import { PlanetaryService } from "@Services";
    // #endregion Imports

    // #region Props
    export let setCurrentImage;
    // #endregion Props

    let keyword = "";
    let notFound = false;

    const showApod = () => {
        keyword = "";
        notFound = false;
        setCurrentImage($photoOfTheDay);
    };

    const makeSearch = async () => {
        if (keyword) {
            if (keyword === $searchTerm && $searchResults.length) {
                showRandomResult($searchResults);
            } else {
                searchTerm.set(keyword);
                PlanetaryService.Search(keyword)
                    .then(result => {
                        const items = result.collection.items.filter(
                            i => i.data[0].media_type === "image",
                        );

                        if (items.length) {
                            notFound = false;
                            showRandomResult(items);
                            searchResults.set(items);
                        } else {
                            notFound = true;
                        }
                    })
                    .catch(e => console.log("err: ", e));
            }
        } else {
            showApod();
        }
    };

    const showRandomResult = items => {
        const randomIndex = Math.round(Math.random() * items.length);
        const randomItem = items[randomIndex];
        const { photographer, secondary_creator, title } = randomItem.data[0];
        const image = {
            url: randomItem.links[0].href,
            title,
            copyright: photographer || secondary_creator || "",
            type: "search",
        };

        setCurrentImage(image);
    };
</script>

<style lang="scss">
    .planetary {
        margin-top: 70px;
        width: 315px;
        display: flex;
        flex-flow: column;
        justify-content: center;
        align-items: center;
    }

    .description {
        width: 100%;
        margin-top: 5px;
        display: flex;
        flex-direction: column;
        align-items: center;
        text-align: center;
    }

    .copyright {
        font-style: italic;
        font-size: 14px;
    }

    .form {
        width: 100%;
        margin-bottom: 15px;
        display: flex;
        flex-direction: column;

        input {
            width: 100%;
            padding-left: 5px;
            margin-right: 10px;
            border-radius: 5px;
        }

        div {
            width: 100%;
            display: flex;
        }
        div:first-child {
            justify-content: space-between;
            margin-bottom: 10px;
        }

        div:nth-child(2) {
            button {
                width: 100%;
            }
        }

        button {
            height: 23px;
            background-color: #f9da2e;
            cursor: pointer;
            border: none;
            border-radius: 5px;
            font-size: 14px;
            padding: 3px 6px;
            font-weight: 700;
            color: #2c3e50;
        }
    }
</style>

<div class="planetary">
    <div class="form">
        <div>
            <input type="text" bind:value={keyword} data-cy="SearchInput" />
            <button on:click={makeSearch} data-cy="SearchButton">Search</button>
        </div>
        <div>
            <button on:click={showApod} data-cy="ApodButton">Photo of the Day</button>
        </div>
    </div>
    {#if notFound}
        <div>No results found for "{keyword}"</div>
    {:else if $imageUrl}
        <img alt={$imageTitle} src={$imageUrl} data-cy="PlanetaryImage" data-type={$imageType} />
        <div class="description">
            {$imageTitle}
            <span class="copyright">{$copyright}</span>
        </div>
    {/if}
</div>
