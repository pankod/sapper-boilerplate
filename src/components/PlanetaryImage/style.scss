.planetary {
    font-family: Arial, Helvetica, sans-serif;
    margin-top: 70px;
    width: 315px;
    display: flex;
    flex-flow: column;
    justify-content: center;
    align-items: center;

    &__form {
        width: 100%;
        margin-bottom: 15px;
        display: flex;
        flex-direction: column;
    
        input {
            width: 100%;
            padding-left: 5px;
            margin-right: 10px;
            border-radius: 5px;
        }
    
        div {
            width: 100%;
            display: flex;
        }

        div:first-child {
            justify-content: space-between;
            margin-bottom: 10px;
        }
    
        div:nth-child(2) {
            button {
                width: 100%;
            }
        }
    
        button {
            height: 23px;
            background-color: #f9da2e;
            cursor: pointer;
            border: none;
            border-radius: 5px;
            font-size: 14px;
            padding: 3px 6px;
            font-weight: 700;
            color: #2c3e50;
        }
    }

    img {
        width: 315px;
    }

    &__description {
        width: 100%;
        margin-top: 5px;
        display: flex;
        flex-direction: column;
        align-items: center;
        text-align: center;
        color: white;

        &__copyright {
            font-style: italic;
            font-size: 14px;
        }
    }
}






