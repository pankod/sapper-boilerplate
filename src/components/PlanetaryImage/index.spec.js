import PlanetaryImage from './index.svelte';
import { render } from '@testing-library/svelte'

const renderComponent = () => {
    return render(PlanetaryImage, {
        setCurrentImage: () => {
            console.log("set current image");
        }
    });
}

describe('PlanetaryImage', () => {
    it("matches snapshot", () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });
    
    it('renders input', () => {
        const { container } = renderComponent();
        expect(container.querySelectorAll("input").length).toEqual(1);
    });
    
    it('renders "Search" button', () => {
        const { container, getAllByText } = renderComponent();
        expect(container.querySelectorAll("button[data-cy='SearchButton']").length).toEqual(1);
        expect(getAllByText("Search").length).toEqual(1);
    });

    it('renders "Photo of the Day" button', () => {
        const { container, getAllByText } = renderComponent();
        expect(container.querySelectorAll("button[data-cy='ApodButton']").length).toEqual(1);
        expect(getAllByText("Photo of the Day").length).toEqual(1);
    });
});
