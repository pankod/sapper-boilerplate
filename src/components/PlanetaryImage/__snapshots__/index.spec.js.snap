// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetaryImage matches snapshot 1`] = `
<body>
  <div>
    <div
      class="planetary"
    >
      <div
        class="planetary__form"
      >
        <div>
          <input
            data-cy="SearchInput"
            type="text"
          />
           
          <button
            data-cy="SearchButton"
          >
            Search
          </button>
        </div>
         
        <div>
          <button
            data-cy="ApodButton"
          >
            Photo of the Day
          </button>
        </div>
      </div>
       
    </div>
  </div>
</body>
`;
